### DISCUSSION TITLE
Select on infinite query can't return a different type

### DISCUSSION THREAD

**Discussion by agos** [TanStack/query#1410] [Ideas] [answered]
**Describe the bug**
It appears that using `select` on an `InfiniteQuery` is limited in potential if using Typescript, as the return type of the select function can't be different from the structure of the passed parameter. This would be useful, for instance to `flatMap` the pages onto a single list - in issue #1187 it was hinted so ("it should be easy to map/filter/flatten here")

**To Reproduce**
Steps to reproduce the behaviour:
1. Create an infinite query
2. Add a `select` function
3. Make the select function return something different from `{ pages, pageParams }`
4. If not returning `page` & `pageParams`, type check will fail. Returning an additional field will work, but trying to use it will make type checking fail.

**Expected behavior**
It should be either possible to
1. return additional fields and have them accessible where you use the Query
2. return a completely different structure from the `select` function

**Comment by boschni** [comment]
A flatmap should be possible like this:
```js
useInfiniteQuery('key', fetchPages, {
  select: data => ({
    pages: data.pages.flatMap(x => x),
    pageParams: data.pageParams
  })
})
```

**Comment by agos** [comment]
@boschni that works for flattening, but I have to keep the `pages` name even if it's not descriptive anymore. It's also not possible to return anything else than an array, so it's still not possible to add more fields

**Comment by boschni** [comment]
I don't think it would be easy to implement something like that but PRs are welcome!

**Comment by AndreyChernykh** [comment]
I've recently had the same issue. In JavaScript I have no problem doing this: 

```js
useInfiniteQuery('key', fetchPages, {
  select: data => ({
    return data.pages.flatMap(x => x)
  })
})
```
In TypeScript we had to use `useMemo` as a clunky alternative to `select` to achieve the similar result. 

The problem is that `select` wraps anything that is returned from there into `InfiniteData<T>`, and there is no way to override this type (at least I couldn't find one) 

https://github.com/tannerlinsley/react-query/blob/52854795b62946b4a4b73e4bddc6954db77e2058/src/core/types.ts#L225

not sure how to fix that properly, though

@tannerlinsley cc

**Reply by TkDodo** [reply]
There should be a way to get this done on type level as it works at runtime, too, but the last time I attempted it, I didn't get very far :/

**Comment by TkDodo** [comment, marked as answer]
we're now tracking this here: https://github.com/tannerlinsley/react-query/issues/3065
